pyOCD/target/memory_map.py

```diff
@@ -0,0 +1,170 @@
+
+    @property
+    def isBootMemory(self):
+        return self._is_boot_mem
+
+    @property
+    def isPoweredOnBoot(self):
+        return self._isPoweredOnBoot
+
+    def containsAddress(self, address):
+        return (address >= self.start) and (address <= self.end)
+
+    def containsRange(self, start, end=None, length=None):
+        if end is None:
+            end = start + length - 1
+        return self.containsAddress(start) and self.containsAddress(end)
+
+    def __str__(self):
+        return "<%s@0x%x name=%s type=%s start=0x%x end=0x%x length=0x%x blocksize=0x%x>" % (self.__class__.__name__, id(self), self.name, self.type, self.start, self.end, self.length, self.blocksize)
+
+    def __repr__(self):
+        return str(self)
+
+## @brief Contiguous region of RAM.
+class RamRegion(MemoryRegion):
+    def __init__(self, start=0, end=0, length=0, blocksize=0, name='', isBootMemory=False, isPoweredOnBoot=True):
+        super(RamRegion, self).__init__(type='ram', start=start, end=end, length=length, name=name, isBootMemory=isBootMemory, isPoweredOnBoot=isPoweredOnBoot)
+
+## @brief Contiguous region of ROM.
+class RomRegion(MemoryRegion):
+    def __init__(self, start=0, end=0, length=0, blocksize=0, name='', isBootMemory=False, isPoweredOnBoot=True):
+        super(RomRegion, self).__init__(type='rom', start=start, end=end, length=length, name=name, isBootMemory=isBootMemory, isPoweredOnBoot=isPoweredOnBoot)
+
+## @brief Contiguous region of flash memory.
+class FlashRegion(MemoryRegion):
+    def __init__(self, start=0, end=0, length=0, blocksize=0, name='', isBootMemory=False, isPoweredOnBoot=True):
+        super(FlashRegion, self).__init__(type='flash', start=start, end=end, length=length, blocksize=blocksize, name=name, isBootMemory=isBootMemory, isPoweredOnBoot=isPoweredOnBoot)
+
+## @brief Memory map consisting of memory regions.
+class MemoryMap(object):
+    def __init__(self, *moreRegions):
+        self._regions = []
+        if len(moreRegions):
+            if type(moreRegions[0]) is list:
+                self._regions = moreRegions[0]
+            else:
+                self._regions.extend(moreRegions)
+        self._regions.sort(key=lambda x:x.start)
+
+    @property
+    def regions(self):
+        return self._regions
+
+    @property
+    def regionCount(self):
+        return len(self._regions)
+
+    def addRegion(self, newRegion):
+        self._regions.append(newRegion)
+        self._regions.sort(key=lambda x:x.start)
+
+    def getBootMemory(self):
+        for r in self._regions:
+            if r.isBootMemory:
+                return r
+        return None
+
+    def getRegionForAddress(self, address):
+        for r in self._regions:
+            if r.containsAddress(address):
+                return r
+        return None
+
+    def isValidAddress(self, address):
+        return self.regionForAddress(address) is not None
+
+    ## @brief Generate GDB memory map XML.
+    def getXML(self):
```

Comment:
Out of curiousity, why not move XML generation into a function that acts on MemoryMap objects in the gdbserver?

It seems like the XML is a very gdb specific thing and this would be the perfect pull request to move it out of the target directory completely.